### ISSUE TITLE
Can a platform object’s associated Realm really change?

### ISSUE THREAD

**Issue by bathos** [whatwg/webidl#1189]
I’m not sure if the following passage is accurate:

> The [Realm](https://tc39.es/ecma262/#realm) that a given [platform object](https://webidl.spec.whatwg.org/#dfn-platform-object) is associated with can change after it has been created. When the [Realm](https://tc39.es/ecma262/#realm) associated with a platform object is changed, its [[Prototype]] [internal slot](https://tc39.es/ecma262/#sec-object-internal-methods-and-internal-slots) must be immediately updated to be the [interface prototype object](https://webidl.spec.whatwg.org/#dfn-interface-prototype-object) of the [primary interface](https://webidl.spec.whatwg.org/#dfn-primary-interface) from the [platform object](https://webidl.spec.whatwg.org/#dfn-platform-object)'s newly associated [Realm](https://tc39.es/ecma262/#realm).

(Links: [bikeshed source](https://github.com/whatwg/webidl/blob/b03aa9e0ceba067f2310431a4ac06f74b3f549c2/index.bs#L13470-L13476); [rendered spec](https://webidl.spec.whatwg.org/#dfn-change-global-environment)).

Supposedly the [associated realm](https://webidl.spec.whatwg.org/#dfn-associated-realm) of a [platform object](https://webidl.spec.whatwg.org/#dfn-platform-object) is, via [relevant realm](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-realm) from HTML, the values of their [[Realm]] slots. Their [[Realm]] slots are allocated and set in steps 4 and 5 of [_internally create a new object implementing the interface_](https://webidl.spec.whatwg.org/#internally-create-a-new-object-implementing-the-interface).

Nothing in Web IDL appears to reassign those values. There’s an exported definition (`dfn-change-global-environment`), but it [doesn’t seem to be referenced](https://github.com/w3c/webref/search?q=dfn-change-global-environment) in other specs, either.

It looks like this was written [9 years ago](https://github.com/whatwg/webidl/commit/2a2ed06099aa5738a07b734276d78e96289d1f94) and hasn’t changed apart from markup/formatting and replacement of “global environment” with “Realm”. Is it still accurate?

My best guess is that it might have been intended to explain the behavior of (what we now call) WindowProxy, which I think [hadn’t yet been formalized](https://blog.whatwg.org/windowproxy-window-and-location). A WindowProxy isn’t itself a platform object and the hypothetical Window objects that they proxy have static [[Realm]] values, as far as I know, so it could just be text that became outdated when that more complete explanation of “windowness” was specified.

**Comment by TimothyGu** [comment]
Historically, I believe Gecko implemented `document.open` by creating a new realm and then "reparenting" existing DOM node objects to the new realm. That logic is since long gone, and assuming there isn't anywhere else that could change the realm of an object (I cannot recall any) we should be able to remove it.

**Comment by petervanderbeken** [comment]
`Document.adoptNode` also reparents in Gecko. I don't know of any plans to change that at this point, though we never got cross-browser agreement on it (eg see also the mega-thread at https://www.w3.org/Bugs/Public/show_bug.cgi?id=20567).

**Comment by bathos** [comment]
@TimothyGu Oh wow — I (think I) remember reading about that [[Prototype]]-swap-on-adoption behavior a long time ago (in spec text I think? was it specified at one point?). It looks like maybe this [impressive/terrifying](https://bugzilla.mozilla.org/show_bug.cgi?id=1489308) set of changes (part 5 & 6?) were where the long gone logic you mentioned got gone. Didn’t realize document.open used to be even weirder!

@petervanderbeken That thread is wild lol. “Re-parenting primitives”! Calling web devs “muggles” unironically!? [This?????](https://www.w3.org/Bugs/Public/show_bug.cgi?id=20567#c12) Only one third through but I love the chaos density & the insight into how folks were thinking about this a decade ago so much — thank you for linking this.

**Comment by domenic** [comment]
Yeah, so we should definitely remove this from Web IDL; great catch. Ideally we'd do so while also committing (or checking for the preexistence of) web platform tests, for the `document.open` and node-adoption cases.
